v2/devdoc/logging.md

```diff
@@ -0,0 +1,162 @@
+
+Defining the sinks to use shall be done by defining a NULL terminated array, where each entry shall be a pointer to a structure that represents the sink interface.
+
+Example:
+
+```c
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+```
+
+### ETW sink
+
+An `ETW` sink shall be supported so that events are emitted via TraceLogging.
+
+For the `ETW` sink there should be a posibility of specifying the provider to use.
+
+### console sink
+
+A console sink shall be supported so that events are printed to the console using `printf`.
+
+## API
+
+```c
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* format, ...);
+```
+
+`log_context` may be `NULL`. If `log_context` is `NULL`, no context information shall be added to the log line.
+
+`format` shall be a `printf`-like format string.
+
+`...` shall be the variable number of arguments required by `format`.
+
+## Contextual logging
+
+For each log line the library shall support specifying a context to be added to the log line.
+
+The context shall allow defining pairs of properties that are added to the event when the event is emitted.
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOCAL_LOG_CONTEXT_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_FIELD`.
+
+Example:
+
+```c
+    LOCAL_LOG_CONTEXT_DEFINE(log_context, NULL, LOG_CONTEXT_FIELD("property_name", "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow wring code like below while not performing the memory copies until the logging is actially executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, NULL, LOG_CONTEXT_FIELD("correlation_id", "%s", correlation_id));
+
+    if (malloc(42) == NULL)
+    {
+        logger_log(LOG_LEVEL_ERROR, local_context, "malloc failed");
+    }
+}
+```
+
+### Dynamically allocated context creation
+
+The library shall support creating a dynamically allocated context.
+
+```c
+#define LOG_CONTEXT_CREATE(log_context_handle, parent_context, ...) \
+
+#define LOG_CONTEXT_CREATE(log_context_handle) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_FIELD`.
+
+Example:
+
+```c
+    LOG_CONTEXT_CREATE(dynamically_allocated_log_context, NULL, LOG_CONTEXT_FIELD("property_name", "%s", MU_P_OR_NULL(prop_value)));
+    LOG_CONTEXT_DESTROY(dynamically_allocated_log_context);
+```
+
+### Context chaining
+
+It shall be supported to chain contexts (define a context or create a context dynamically while specifying a parent context to inherit the information from).
```

Comment:
>chain [](http://example.com/codeflow?start=25&length=5)

Yeah, this is most awesome.

Would be really nice if we'd have in this document some spelled out example onf "what's on the screen" when logging in a chained context.

Reply:
Added output of example. in the .md